### PR TITLE
update field name

### PR DESCRIPTION
**Pull request by siddontang**
Use as name explicitly. Old code uses as name for field name, if has not, uses origin expression string instead. e,g, `select 1 as a`, we will use "a" as the field name, `select 1`, we will use "1" as the field name. 

This mechanism is not proper for later field name handing in group by, order by or other clauses. e.g, `select t.c as "t.c" from t`, using the old way we cannot know "t.c" is an alia name or just expression representation. So now the `GetField*` function in field package is complex and confused. 

I think we will use AsName in field struct explicitly, we will check AsName to see whether a field has an alias name or not, if no alias name, we can try to use expression representation then.

### CODE REVIEW

File: field/result_field_test.go [hunk 1]
```diff
@@ -144,12 +144,12 @@ func (*testResultFieldSuite) TestMain(c *C) {
 
 	// For GetFieldIndex
 	f1 := &field.Field{
-		Expr: &expression.Ident{CIStr: model.NewCIStr("c1")},
-		Name: "a",
+		Expr:   &expression.Ident{CIStr: model.NewCIStr("c1")},
```
Comment (shenli) [inline field/result_field_test.go]: Why use '&' herer but does not use '&' in the above test file?

Comment (siddontang) [inline field/result_field_test.go]: I don't see any other place not using "&" for expression.Ident.

Comment (shenli) [inline field/result_field_test.go]: Value is kind of weird, Maybe we should change its function to use pointer.  
Expr:   expression.Value{Val: "c1+1"},

**Comment by shenli** [comment]
LGTM

File: field/result_field_test.go (hunk 1 above)
Comment (qiuyesuifeng) [inline field/result_field_test.go]: +1, all expressions use pointer is better.

File: parser/parser.y [hunk 2]
```diff
@@ -1536,10 +1536,7 @@ Field:
 |	Expression Field1
 	{
 		expr, name := expression.Expr($1), $2.(string)
-		if name == "" {
-			name = expr.String()
-		}
-		$$ = &field.Field{Expr: expr, Name: name}
+		$$ = &field.Field{Expr: expr, AsName: name}
```
Comment (qiuyesuifeng) [inline parser/parser.y]: How about
```
$$ = &field.Field{
     Expr: expression.Expr($1),
     AsName: $2.(string)
}
```
seems needless to define variables.

Comment (siddontang) [inline parser/parser.y]: Em, seems no big different. :-)

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**